### ISSUE TITLE
Walrus operator overwrites type hint

### ISSUE THREAD

**Issue by FabiPi3** [python/mypy#17576]
Suppose I have some data coming from an untyped or dynamic source, which I don't know the type. Mypy correctly says it has the `Any` type. Now I want to extract some information, where I know the type. That's why I type hint it. But using the walrus operator, the type hint seems to not being considered. See the example code.

Gist URL: https://gist.github.com/mypy-play/e59d20e8470058c5f2c261338bf796de
Playground URL: https://mypy-play.net/?mypy=latest&python=3.12&gist=e59d20e8470058c5f2c261338bf796de

```python
data: Any = {"a": "abc"}
reveal_type(data)  # Any, which is correct
a: str
if a := data["a"]:
    pass

reveal_type(a)  # mypy: Any, which is weird

c: str
c = data["a"]
if c:
    pass

reveal_type(c)  # without walrus its string!
```

**Expected Behavior**

variable `a` should have type string as well.

**Actual Behavior**

its Any type. Output:

```
main.py:4: note: Revealed type is "Any"
main.py:9: note: Revealed type is "Any"
main.py:16: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file
```

**Your Environment**

python 3.12, mypy 1.11.0

**Comment by ilevkivskyi** [comment]
The two examples behave consistently now (both give `str`).
